torch/nn/modules/linear.py

```diff
@@ -59,5 +59,66 @@ def __repr__(self):
             + str(self.out_features) + ')'
 
 
-# TODO: Bilinear
+class BiLinear(Module):
+    r"""Applies a bilinear transformation to the incoming data: :math:`\forall k: y_k = x_1 A_k x_2 + b`
+
+    Args:
+        in_features_1: size of each input sample of the first input
+        in_features_2: size of each input sample of the second input
+        out_features: size of each output sample
+        bias: If set to False, the layer will not learn an additive bias. Default: True
+
+    Shape:
+        - Input: :math:`[(N, in\_features\_1), (N, in\_features\_2)`
+        - Output: :math:`(N, out\_features)`
+
+    Attributes:
+        weight: the learnable weights of the module of shape (out_features x in_features_1 x in_features_2)
+        bias:   the learnable bias of the module of shape (out_features)
+
+    Examples::
+
+        >>> m = nn.BiLinear(10, 20, 30)
+        >>> input_1 = autograd.Variable(torch.randn(128, 20))
+        >>> input_2 = autograd.Variable(torch.randn(128, 30))
+        >>> output = m([input_1, input_2])
+        >>> print(output.size())
+    """
+
+    def __init__(self, in_features_1, in_features_2, out_features, bias=True):
+        super(BiLinear, self).__init__()
+        self.in_features_1 = in_features_1
+        self.in_features_2 = in_features_2
+        self.out_features = out_features
+        self.weight = Parameter(torch.Tensor(out_features, in_features_1, in_features_2))
+        if bias:
+            self.bias = Parameter(torch.Tensor(out_features))
+        else:
+            self.register_parameter('bias', None)
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        stdv = 1. / math.sqrt(self.weight.size(1))
+        self.weight.data.uniform_(-stdv, stdv)
+        if self.bias is not None:
+            self.bias.data.uniform_(-stdv, stdv)
+
+    def forward(self, input):
+        assert len(input) == 2
+        output = torch.cat([input[0].mm(self.weight[i]).mul(input[1]).sum(1)
+                            for i in range(self.out_features)], 1)
+        if self.bias is not None:
+            output.add_(self.bias.expand_as(output))
```

Comment:
Can you use the `addr_` trick from [`Linear`](https://github.com/pytorch/pytorch/blob/master/torch/nn/_functions/linear.py#L11-L14)? It's a bit faster.